### PR TITLE
RFC-0051: EXCLUDE Clause

### PR DESCRIPTION
**Pull request by alancai98**
*Issue #, if available:* https://github.com/partiql/partiql-lang/issues/27

RFC to define the `EXCLUDE` operator and definition in terms of existing PartiQL operators. Current reference implementation is in partiql-lang-kotlin's `EvaluatingCompiler` (though I'm currently working to port it to the `PhysicalPlanCompilerImpl`).

[Rendered doc](https://github.com/partiql/partiql-docs/blob/exclude-rfc/RFCs/0051-exclude-operator.adoc)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: RFCs/0051-exclude-operator.adoc [hunk 1]
```diff
@@ -0,0 +1,1661 @@
...
+[[anchor-2]]
+.Rewrite rule 2: `EXCLUDE` steps `s~1~,...,s~n-1~`
+For this rewrite rule definition, let `<v~i-1~>` be the identifier created from the previous exclude step (or `r` if this is the first step). For some exclude step `s~i~` that is not the last step, we case on the type of exclude step.
+
+[[anchor-2ai]] Rule 2.a.i::
+    If `s~i~` is a case-sensitive tuple attribute exclude step (e.g. `."foo"` or `['foo']`), where `<v~i~>` and `<attr~i~>` are fresh variables, add the following `WHEN` branch to the `i`^th^ nested `CASE`.
+[source,partiql,subs="+{markup-in-source}"]
+----
+WHEN <v~i-1~> IS STRUCT THEN
+    PIVOT (
```
Comment (alancai98) [inline RFCs/0051-exclude-operator.adoc:199]: Brought up in a separate conversation w/ Josh, may need to surround the `PIVOT ... FROM UNPIVOT ... ` (or any other subquery) in parens for SQL compatibility. I'll add parens to these subqueries.

Comment (alancai98) [reply RFCs/0051-exclude-operator.adoc:199]: Resolved in: https://github.com/partiql/partiql-docs/pull/51/commits/3910333bf0c850efafc177e205cecd94c222f764

File: RFCs/0051-exclude-operator.adoc [hunk 2]
```diff
@@ -0,0 +1,1661 @@
...
+* PartiQL Issue: https://github.com/partiql/partiql-spec/issues/39
+* RFC PR: https://github.com/partiql/partiql-docs/pull/51
+
+== Summary
+
+This doc defines the `EXCLUDE` binding tuple operator used to omit nested values before projection and defines the semantics in terms of syntactic rewrites with existing operators.
+
+== Motivation
+
+SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[slack overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.
```
[the remark this answers is not in the document]
Comment (alancai98) [reply RFCs/0051-exclude-operator.adoc:16]: Ah whoops. I'll fix.

```suggestion
SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[Stack Overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.
```

File: RFCs/0051-exclude-operator.adoc [hunk 3]
```diff
@@ -0,0 +1,1661 @@
...
+
+If there are any `EXCLUDE` paths of length `i`, then similar to <<anchor-3ai, Rule 3.a.i>> and <<anchor-3aii, Rule 3.a.ii>>, we add a `WHERE` clause to filter out those fields. The fields to exclude will be grouped together based on if the tuple attribute exclude step was case sensitive or case-insensitive.
+
+If there are any `EXCLUDE` paths of length greater than `i`, then similar to <<anchor-2ai, Rule 2.a.i>> and <<anchor-2aii, Rule 2.a.ii>>, we add a `CASE` expression within the `PIVOT`. This `CASE` expression within the `PIVOT` will define a `WHEN` branch for each of the unique tuple attribute steps. Each of these `WHEN` branches will apply the rewrite rules for the exclude paths that have additional steps and equivalent tuple attribute or tuple wildcard. An `ELSE` branch will be added to this `CASE` expression which will apply the rewrite rules for the exclude paths with a tuple wildcard at level `i` and additional steps.
+[source,partiql,subs="+{markup-in-source}"]
+----
+-- Let `k` represent the number of unique exclude tuple attrs for paths of length
+-- greater than `i`.
+-- `<v~i~>` and `<attr~i~>` are fresh variables
```
[the remark this answers is not in the document]
Comment (alancai98) [reply RFCs/0051-exclude-operator.adoc:396]: I'll take another look at the variables in these multi-exclude path rewrites. Agree that it can be confusing w/ how it's currently written.

File: RFCs/0051-exclude-operator.adoc [hunk 4]
```diff
@@ -0,0 +1,1700 @@
...
+[quote,https://issues.apache.org/jira/browse/ASTERIXDB-3059]
+____
+'EXCLUDE' (used in lieu of 'EXCEPT' to avoid confusion with the set operation)
+____
++
+Also of the databases sampled that have a similar clause (see <<Prior art>>), more had chosen `EXCLUDE` over `EXCEPT`.
+
+Why is `EXCLUDE` modeled as a binding tuple operator as opposed to a value expression?::
+
+We had also considered modeling `EXCLUDE` as a value operation evaluated after the `<select clause>`. Evaluating `EXCLUDE` last could contradict the PartiQL specification's assertion that the `<select clause>` is evaluated last, which may add confusion. There were also some additional edge cases that complicated defining `EXCLUDE` as a value operator. For example, let's look at the following query:
```
[the remark this answers is not in the document]
Comment (alancai98) [reply RFCs/0051-exclude-operator.adoc:1478]: > Did you explore `EXCLUDE` as a value operation (maybe just a special-form) evaluated during the `<select clause>`? Similar to how any other special-form function invocation would work in the projection list?

Could you clarify what you mean by special-form function? Is this referring to non-standard function syntax (kinda like what substring, overlay do)?

Regarding modeling `EXCLUDE` as a value operation/function evaluated within the `SELECT` clause, we had also explored this option. These aren't quite functions though, since function arguments expect value expressions and exclude paths do not return values.

The above way you had modeled expressions evaluated left-to-right could be problematic. Consider `SELECT EXCLUDE t.a[0], EXCLUDE t.a[1].field FROM ...`. Should the second exclude path remove `field` from the `a`'s original `1` index or after evaluating the first exclude path (i.e. originally the `2` index). We found considering the paths as associative and evaluating the paths in parallel (i.e. applying all exclude paths on the same value/binding) as more intuitive.

**Review by alancai98 — COMMENTED** [review]
Thanks for the review, John. Once I get back after the Thanksgiving holiday, I'll followup to your comments and apply your feedback around variable definitions + `IS TUPLE`/`IS ARRAY` in the next revision.

File: RFCs/0051-exclude-operator.adoc [hunk 5]
```diff
@@ -0,0 +1,1661 @@
...
+        END
+    )
+    AT <attr~i~>
+    FROM UNPIVOT <v~i-1~> AS <v~i~> AT <attr~i~>
+----
+[[anchor-2aii]] Rule 2.a.ii::
+    If `s~i~` is a case-insensitive tuple attribute exclude step (e.g. `.foo`), where `<v~i~>` and `<attr~i~>` are fresh variables, add the following `WHEN` branch to the the `i`^th^ nested `CASE`.
+[source,partiql,subs="+{markup-in-source}"]
+----
+WHEN <v~i-1~> IS STRUCT THEN
```
[the remark this answers is not in the document]
Comment (alancai98) [reply RFCs/0051-exclude-operator.adoc]: (similar to other comment regarding `IS ARRAY`), I'll fix in next revision to use `IS TUPLE`.

**Review by alancai98 — COMMENTED** [review]
New revision addresses John's comments related to
- `IS STRUCT` -> `IS TUPLE`, `IS LIST` -> `IS ARRAY` and related prose
- more consistent variable definitions through the doc
- change in multi-exclude path variable definitions
- additional subsumption rule for empty steps
- typo fixes

File: RFCs/0051-exclude-operator.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnedquinn) [reply RFCs/0051-exclude-operator.adoc:1481]: 1. Yes, meaning non-standard function syntax.
2. Yes -- but that is the AST, which isn't the same as execution. We don't operate on the AST -- we convert `TRIM` to `$trim_leading` and `CURRENT_USER` to `$current_user`. The SQL:1999 Specification even talks about the conversion of the syntactic `BETWEEN` to comparison operators -- the theoretical `between` AST node disappears and gets replaced. We could just as easily say `<lhs> EXCLUDE ( <exclude-arg-1>, <exclude-arg-2> )` gets converted to `EXCLUDE(<lhs>, [ <exclude-arg-1-stringified>, <exclude-arg-2-stringified>])`. Anyways, it *could* provide flexibility, and I believe it's possible to model.
3. We could probably look at BigQuery which expects the RHS to be parenthesized. Ex: `SELECT * EXCEPT (a, b)`. I would imagine this query in BigQuery strips out both `a` and `b` simultaneously.

Given PartiQL's flexibility, I'm personally intrigued by the prospect of writing:
```
SELECT VALUE {
  'full_name': person.l_name || ', ' person.f_name,
  'encrypted_ssn': udf_to_encrypt(person.ssn),
  'all_other_details': person EXCLUDE (f_name, l_name, ssn)
} FROM Persons AS person
```

Especially since tuples are first-class citizens, I find it long overdue for some practical built-ins (beyond `TUPLEUNION` which PLK doesn't have implemented -- yet). It seems like a potential opportunity to allow for its introduction, especially since SQL's `SELECT` gets transformed to PartiQL's `SELECT VALUE <tuple>`.

File: RFCs/0051-exclude-operator.adoc [hunk 6]
```diff
@@ -0,0 +1,1703 @@
...
+
+== Summary
+
+This doc defines the `EXCLUDE` binding tuple operator used to omit nested values before projection and defines the semantics in terms of syntactic rewrites with existing operators.
+
+== Motivation
+
+SQL users often use `SELECT *` to project all of the columns of a table. There is frequently a use case in which a user would like to project all the columns from a table other than a subset of the columns (see https://stackoverflow.com/q/729197[Stack Overflow question]). There are workarounds in some database systems that are somewhat inefficient (e.g. creating a new table and dropping a specific column), but it can be helpful to have a dedicated syntax to filter out certain columns. <<Prior art>> lists out a few databases that provide some version of this column filtering.
+
+There is a similar need among PartiQL users to exclude certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:18]:
```suggestion
There is a similar need among PartiQL users to exclude certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields other than the field(s) that need to get excluded or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.
```

File: RFCs/0051-exclude-operator.adoc [hunk 7]
```diff
@@ -0,0 +1,1703 @@
...
+
+<identifier> ::= (* See identifier in PartiQL spec figure 4 *)
+
+<asterisk> ::= "*"
+<dot> :: = "."
+<left bracket> ::= "["
+<right bracket> ::= "]"
+----
+
+NOTE: Despite their similar syntax and naming, ``<exclude path>``s are different than PartiQL path expressions.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:67]: Text can also include the spec section and link for Path expressions.

File: RFCs/0051-exclude-operator.adoc [hunk 8]
```diff
@@ -0,0 +1,1703 @@
...
+** `[*]` - collection wildcard exclude step
+
+=== Out of scope / assumptions
+
+* We restrict tuple attribute exclude steps to use string literals and collection index exclude steps to use int literals. Thus `<exclude paths>` are statically known. We can decide whether to add other exclude paths (e.g. expressions) if a use case arises.
+* If sufficient schema is present and the path can be resolved, we assume the root of an `EXCLUDE` path can be omitted. The variable resolution rules follow what is already included in the PartiQL specification.
+* We require that every fully-qualified `<exclude path>` contain a root and at least one step. If a use case arises to exclude a binding tuple variable, then this functionality can be added.
+* S-expressions are part of the Ion type system.footnote:[https://amazon-ion.github.io/ion-docs/docs/spec.html#sexp]
+PartiQL should support s-expression types and values since PartiQL's type system is a superset over the Ion types. Because the current PartiQL specification does not formally define s-expressions operations, we consider the definition of collection index and wildcard steps on s-expressions as out-of-scope for this RFC.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:92]: Perhaps the statement can be less assertive; I know this is one of those hotly debated topics. The spec. says:
> PartiQL’s data model extends SQL to Ion’s type system to cover schema-less and nested data. Such values can be
directly quoted with `quotes.

So text can just convey the message that s-expressions semantics as a collection type is not fully defined yet, hence is out of the scope.

File: RFCs/0051-exclude-operator.adoc [hunk 9]
```diff
@@ -0,0 +1,1703 @@
...
+
+=== Rewrite Procedure
+==== Step 1: subsumption of `EXCLUDE` paths
+We perform the following step to ensure that there are no redundant `EXCLUDE` paths. That is, there is no path such that all of its excluded binding tuple values are excluded by another exclude path.footnote:[This subsumption step is included to make the subsequent rewrite steps easier to reason about. In a query without redundant exclude paths, this step is not necessary.]
+
+For each `<exclude path>` `p=root~p~s~1~...s~x~`, we compare it with all other ``<exclude path>``s. `<exclude path>` `p` is said to be subsumed by another path `q=root~q~t~1~...t~y~` and not included in the rewritten `EXCLUDE` clause if any of the following rules apply:
+
+NOTE: The following rules assume `root~p~=root~q~`.
+
+.Subsumption rules
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:102]: I know we have the `table 1.` for the examples, but adding an example for each of the rules would also enhance readability.

File: RFCs/0051-exclude-operator.adoc [hunk 10]
```diff
@@ -0,0 +1,1703 @@
...
+[[anchor-1a]] Rule 1.a::
+    If `y = 0` (i.e. `q` has no steps), `q` subsumes `p`.
+[[anchor-1b]] Rule 1.b::
+    If `x ≥ y` and `s~1~...s~x~=t~1~...t~x~`, `q` subsumes `p`. Put another way if `p` has at least as many steps as `q` and the steps up to ``q``'s length are equivalent, `q` subsumes `p`.
+
+Otherwise, there must be some step at which `p` and `q` diverge. Let's call this step's index `i`.
+
+[[anchor-1c]] Rule 1.c::
+    If `s~i~` is a tuple attribute and `t~i~` is a tuple wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:111]: For this and others that apply: In order for the underlying subsumption to happen (I.e., `t~i+1~...t~y~` subsumes `s~i+1~...s~x~`) the two should be considered as independent `<exclude path>`s so that they can have root; is that so? if yes, text may need to clarify that further.

Comment (am357) [inline RFCs/0051-exclude-operator.adoc:111]:
```suggestion
    Given `i < y and i < x`, if `s~i~` is a tuple attribute and `t~i~` is a tuple wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

File: RFCs/0051-exclude-operator.adoc [hunk 11]
```diff
@@ -0,0 +1,1703 @@
...
+[[anchor-1d]] Rule 1.d::
+    If `s~i~` is a collection index and `t~i~` is a collection wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:113]:
```suggestion
    Given `i < y and i < x`, if `s~i~` is a collection index and `t~i~` is a collection wildcard and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

File: RFCs/0051-exclude-operator.adoc [hunk 12]
```diff
@@ -0,0 +1,1703 @@
...
+[[anchor-1e]] Rule 1.e::
+    If `s~i~` is a case-sensitive tuple attribute and `t~i~` is a case-insensitive tuple attribute and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:115]:
```suggestion
    Given `i < y and i < x`, if `s~i~` is a case-sensitive tuple attribute and `t~i~` is a case-insensitive tuple attribute and `t~i+1~...t~y~` subsumes `s~i+1~...s~x~` (i.e. the steps following `t~i~` subsumes the steps following `s~i~`), then `q` subsumes `p`.
```

File: RFCs/0051-exclude-operator.adoc [hunk 13]
```diff
@@ -0,0 +1,1703 @@
...
+|`t.a.b.c`    |`t.a.b`     |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
+|`t.a.b.c`    |`t.a.b.*`   |`q` subsumes `p` (by <<anchor-1c, 1.c>> then <<anchor-1a, 1.a>>)
+|`t.a.b.c`    |`t.a.*.c`   |`q` subsumes `p` (by <<anchor-1c, 1.c>> then <<anchor-1b, 1.b>>)
+|`t.a.b[1]`   |`t.a.b`     |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
+|`t.a.b[1]`   |`t.a.b[*]`  |`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1a, 1.a>>)
+|`t.a.b[1].c` |`t.a.b[1]`  |`q` subsumes `p` (by <<anchor-1b, 1.b>>)
+|`t.a.b[1].c` |`t.a.b[*]`  |`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1a, 1.a>>)
+|`t.a.b[1].c` |`t.a.b[*].c`|`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1b, 1.b>>)
+|`t.a."b"`    |`t.a.b`     |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1a, 1.a>>)
+|`t.a."b".c`  |`t.a.b.c`   |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1b, 1.b>>)
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:135]: Would be good to have a 'No subsumption rule apply` for case sensitive mismatch.

File: RFCs/0051-exclude-operator.adoc [hunk 14]
```diff
@@ -0,0 +1,1703 @@
...
+<select clause>
+EXCLUDE r.s~1~...s~n~
+<from clause>
+<other clauses>
+
+-- Rewritten to:
+<select clause>
+FROM (
+    SELECT VALUE {
+        'r': -- Apply below rewrite rules for steps `s~1~...s~n~`
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:159]:
```suggestion
        'r': -- Apply rewrite rules explained in the following sections for steps `s~1~...s~n~`
```

File: RFCs/0051-exclude-operator.adoc [hunk 15]
```diff
@@ -0,0 +1,1703 @@
...
+        ...  -- Other vars created from the other clauses
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:160]:
```suggestion
        ...  -- Add other variables created from the other clauses using identity function
```

File: RFCs/0051-exclude-operator.adoc [hunk 16]
```diff
@@ -0,0 +1,1703 @@
...
+    AS <topLevelTbl> AT <idx>
+    ORDER BY <idx>
+]
+----
+
+
+The main idea for rewriting the `EXCLUDE` steps `s~1~,...,s~n~` is to create a nested `CASE` expression for each step, whereby the nested `CASE` expressions for `s~1~,...,s~n-1~` unnest the input binding tuple and the final `CASE` expression for `s~n~` (i.e. the final step) filters out the desired tuple field(s) or collection index(es). Every exclude step has an expected type to process during evaluation. Tuple attribute and wildcard exclude steps expect a tuple. Whereas a collection index expects an array and a collection wildcard expects an array or bag. The `CASE` expression at each level `i` recreates this expected type by including a `WHEN` branch based on the expected type. Each `CASE` expression will include an `ELSE` branch which outputs the previous level's identifier. This set of branches ensures that at evaluation time, if there is a type mismatch (e.g. evaluation value is an array while the exclude step is a tuple attribute), there is no evaluation error and the previous level's value is returned through the `ELSE` branch. This behavior applies to both the permissive and strict typing modes.
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:173]:
```suggestion
The main idea for rewriting the `EXCLUDE` steps `s~1~,...,s~n~` is to create a nested `CASE` expression for each step, whereby the nested `CASE` expressions for `s~1~,...,s~n-1~` unnest the input binding tuple and the final `CASE` expression for `s~n~` (i.e. the final step) filters out the desired tuple field(s) or collection index(es). Every exclude step has an expected type to process during evaluation. Tuple attribute and wildcard exclude steps expect a `Tuple`, whereas a collection index expects `Array` and a collection wildcard expects `Array` or `Bag` types. The `CASE` expression at each level `i` recreates this expected type by including a `WHEN` branch based on the expected type. Each `CASE` expression will include an `ELSE` branch which outputs the previous level's identifier. This set of branches ensures that at evaluation time, if there is a type mismatch (e.g. evaluation value is an array while the exclude step is a tuple attribute), there is no evaluation error and the previous level's value is returned through the `ELSE` branch. This behavior applies to both the permissive and strict typing modes.
```

File: RFCs/0051-exclude-operator.adoc [hunk 17]
```diff
@@ -0,0 +1,1703 @@
...
+:markup-in-source: verbatim,quotes
+
+
+* Start Date: 2023-11-07
+* PartiQL Issue: https://github.com/partiql/partiql-lang/issues/27
+* RFC PR: https://github.com/partiql/partiql-docs/pull/51
+
+== Summary
+
+This doc defines the `EXCLUDE` binding tuple operator used to omit nested values before projection and defines the semantics in terms of syntactic rewrites with existing operators.
```
Comment (jpschorr) [inline RFCs/0051-exclude-operator.adoc:12]:
```suggestion
This doc defines the `EXCLUDE` binding tuple operator used to omit (potentially) nested values before projection and defines the semantics in terms of syntactic rewrites into existing operators.
```

File: RFCs/0051-exclude-operator.adoc (hunk 6 above)
Comment (jpschorr) [inline RFCs/0051-exclude-operator.adoc:18]:
```suggestion
There is a similar need among PartiQL users to exclude not only top-level, but also certain nested fields from semi-structured data. PartiQL supports `SELECT *` to project all of the fields of a binding tuple. If a user wanted to omit one field from this projection, they would need to list out all of the projection fields or perform some intricate combination of `PIVOT` and ``UNPIVOT``s.
```

**Review by am357 — COMMENTED** [review]
I still need to review the rest of RFC; adding the portion that I have reviewed so far.

File: RFCs/0051-exclude-operator.adoc (hunk 8 above)
Comment (jpschorr) [reply RFCs/0051-exclude-operator.adoc:92]: Agreed. This statement makes more assertions about the PartiQL value system than does the spec.

File: RFCs/0051-exclude-operator.adoc [hunk 18]
```diff
@@ -0,0 +1,1703 @@
...
+*** `.<quoted identifier>`/`[<string literal>]` - case-_sensitive_ tuple attribute
+*** `.<unquoted identifier>` - case-_insensitive_ tuple attribute
+** `[<int literal>]` - collection index exclude step
+** `.*` - tuple wildcard exclude step
+** `[*]` - collection wildcard exclude step
+
+=== Out of scope / assumptions
+
+* We restrict tuple attribute exclude steps to use string literals and collection index exclude steps to use int literals. Thus `<exclude paths>` are statically known. We can decide whether to add other exclude paths (e.g. expressions) if a use case arises.
+* If sufficient schema is present and the path can be resolved, we assume the root of an `EXCLUDE` path can be omitted. The variable resolution rules follow what is already included in the PartiQL specification.
```
Comment (almann) [inline RFCs/0051-exclude-operator.adoc:89]: I think we might want to have an example of attribute as a variable.

File: RFCs/0051-exclude-operator.adoc
Comment (almann) [inline RFCs/0051-exclude-operator.adoc:1]: Top-level, let's format these lines to be like 80 or 120 characters wide.

File: RFCs/0051-exclude-operator.adoc [hunk 19]
```diff
@@ -0,0 +1,1703 @@
...
+* We require that every fully-qualified `<exclude path>` contain a root and at least one step. If a use case arises to exclude a binding tuple variable, then this functionality can be added.
```
Comment (almann) [inline RFCs/0051-exclude-operator.adoc:90]: What is the rationale for this limitation?  We should put that here.

File: RFCs/0051-exclude-operator.adoc [hunk 20]
```diff
@@ -0,0 +1,1703 @@
...
+|`t.a.b[1].c` |`t.a.b[*]`  |`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1a, 1.a>>)
+|`t.a.b[1].c` |`t.a.b[*].c`|`q` subsumes `p` (by <<anchor-1d, 1.d>> then <<anchor-1b, 1.b>>)
+|`t.a."b"`    |`t.a.b`     |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1a, 1.a>>)
+|`t.a."b".c`  |`t.a.b.c`   |`q` subsumes `p` (by <<anchor-1e, 1.e>> then <<anchor-1b, 1.b>>)
+|=======================
+
+---
+We first illustrate the rewrite rule for a single `EXCLUDE` path and then explain the syntax rewrite for multiple exclude paths.
+
+==== Step 2 (single): rewrite a single `EXCLUDE` path
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:141]: Considering that there are many rules, perhaps a pseudo-code accompanying the text explanation could also be helpful.

File: RFCs/0051-exclude-operator.adoc (hunk 8 above)
Comment (jpschorr) [inline RFCs/0051-exclude-operator.adoc:92]: I would reword this section. I would avoid usage of terms like 'we'  and write it more formally and more tersely.

e.g.,

```
=== Limitations

* This RFC requires that every fully-qualified `<exclude path>` contain a root and at least one step.
* This RFC restricts tuple attribute exclude steps to use string literals and collection index exclude steps to use int literals. Thus `<exclude paths>` are statically known. This 
* This RFC makes no changes to schema and name inference, and assumes that such inference is run as a prerequisite.
* This RFC defines `<exclude paths>` only over `list`, `bag`, and `tuple` value collections.

```

File: RFCs/0051-exclude-operator.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (almann) [reply RFCs/0051-exclude-operator.adoc:1481]: Cannot be a value operator if you want to ever prune binding tuple variables.

File: RFCs/0051-exclude-operator.adoc [hunk 21]
```diff
@@ -0,0 +1,1703 @@
...
+* More info on MongoDB: https://dbdb.io/db/mongodb
+
+MongoDB supports excluding certain fields by setting the attribute's value to `0`. The following returns all fields other than the `status` and `instock` field for all of the matching documents.
+
+[source,shell]
+----
+db.inventory.find( { status: "A" }, { status: 0, instock: 0 } )
+----
+
+== Unresolved questions
```
Comment (am357) [inline RFCs/0051-exclude-operator.adoc:1663]: Wondering if partial schema case be out of scope only? i.e. having SQL's schema-full within the scope as well.

**Comment by johnedquinn** [comment]
I have come across a scenario where I'd like to express caution. It has to do with scoping of variables specifically when `EXCLUDE` is present. Consider the following simple query

```sql
SELECT
    t.a,
    t.b -- this won't produce anything!
EXCLUDE t.b
FROM t
```

This makes sense. However, I've been dealing with nested queries, and I'm curious what would happen in the following scenario:
```sql
SELECT
    (
        SELECT t2.c + t1.a -- this shouldn't work!
        EXCLUDE t1.a
        FROM t2
    ) AS t1_plus_t2
FROM t1 -- this has columns a and b
```

In the above, your RFC works great I believe. The query should fail. However, this *may* cause problems if `EXCLUDE` is allowed to remove bindings entirely. In my opinion, here is the query (again) with some more information regarding binding environments:
```sql
global env = E0 = < t1: <<{a, b}>>, t2: <<{c, d}>> >
SELECT
    (
        SELECT t2.c + t1.a AS x-- input env = E0 || E1 || E2 = < t1: {b}, t2: {c, d} >. Output env = < x >
        EXCLUDE t1.a -- input env = E0 || E1 || E2. Output Env = E0 || E1 || E2 (with some minor eliminations of attributes) = < t1: {b}, t2: {c, d} >
        FROM t2 -- input env = E0 || E1. Output env (E2) = E0 || E1 || < t2: {c, d} > = < t1: {a, b}, t2: {c, d} >
    ) AS t1_plus_t2 -- the whole SELECT (including this projection item subquery) has input env = E0 || E1
FROM t1 -- input env = E0, output env (E1) = E0 || < t1: { a, b } > = < t1: {a, b}, t2: <<{c, d}>> >
```

If we want to make sure that the inner select does NOT get access to what is being excluded, then we must not allow EXCLUDE to exclude whole bindings (rather than attributes of bindings). If we had a very similar query that excluded an entire binding, we might still allow the projection list to access the original `t1.a`. See below:

```sql
global env = E0 = < t1: <<{a, b}>>, t2: <<{c, d}>> >
SELECT
    (
        SELECT t2.c + t1.a AS x-- input env = E0 || E1 || E3 = < t1: {a, b}, t2: {c, d} >. Output env (E4) = < x >
        EXCLUDE t1 -- input env = E0 || E1 || E2. Output Env (E3) = E0 || E1 || E2 = < t2: {c, d} >
        FROM t2 -- input env = E0 || E1. Output env (E2) = E0 || E1 || < t2: {c, d} > = < t1: {a, b}, t2: {c, d} >
    ) AS t1_plus_t2 -- the whole SELECT (including this projection item subquery) has input env = E0 || E1
FROM t1 -- input env = E0, output env (E1) = E0 || < t1: { a, b } > = < t1: {a, b}, t2: <<{c, d}>> >
```

Notice that the inner select still received `t1` in its entirety due to the concatenation of environments (effectively bypassing the exclusion). One might ask: "That wouldn't be the case. Bindings always come from their inputs. Why is this different?" We might look at another operator that strips out bindings, the aggregation. In other engines, we can still access the outer variables. See the following examples:
- [SQLite](https://www.db-fiddle.com/f/knBFPpHmdGQbyM8LLWdR5P/0)
- [Postgres](https://www.db-fiddle.com/f/knBFPpHmdGQbyM8LLWdR5P/1)
- [MySQL](https://www.db-fiddle.com/f/knBFPpHmdGQbyM8LLWdR5P/2)
